logstash-core/src/main/java/org/logstash/config/ir/CompiledPipeline.java

```diff
@@ -243,27 +243,42 @@ private Map<String, Object> expandArguments(final PluginDefinition pluginDefinit
     }
 
     @SuppressWarnings({"rawtypes", "unchecked"})
-    public static Map<String, Object> expandConfigVariables(ConfigVariableExpander cve, Map<String, Object> configArgs) {
+    public static Map<String, Object> expandConfigVariables(ConfigVariableExpander cve, Map<String, Object> configArgs, boolean keepSecrets) {
```

Comment:
is there a situation where we don't want to `keepSecrets` to be true?

Reply:
Yes, in the expansion of arguments:
```java
    private Map<String, Object> expandArguments(final PluginDefinition pluginDefinition, final ConfigVariableExpander cve) {
        Map<String, Object> arguments = expandConfigVariables(cve, pluginDefinition.getArguments());
```

This expansion is done when the plugin instances are created. If a pipeline config contains a plugin option captured from `SecretStore` than that argument has to be expanded before creating the plugin, else the plugin sees an instance of `Password` instead of `String` (like it happens today).

Instead, in `Expression` evaluation the expansion of the secret has to be done during the evaluation, or better, when the the expression is compiled.